pkg/controllers/disruption/helpers.go

```diff
@@ -282,6 +225,64 @@ func BuildNodePoolMap(ctx context.Context, kubeClient client.Client, cloudProvid
+// We calculate allowed disruptions by taking the max disruptions allowed by disruption reason and subtracting the number of nodes that are already being deleted by that disruption reason.
+//
+//nolint:gocyclo
+func BuildDisruptionBudgets(ctx context.Context, cluster *state.Cluster, clk clock.Clock, kubeClient client.Client, recorder events.Recorder) (map[string]map[v1beta1.DisruptionReason]int, error) {
+	disruptionBudgetMapping := map[string]map[v1beta1.DisruptionReason]int{}
+	numNodes := map[string]int{}   // map[nodepool] -> node count in nodepool
+	disrupting := map[string]int{} // map[nodepool] -> nodes undergoing disruption
+	for _, node := range cluster.Nodes() {
+		// We only consider nodes that we own and are initialized towards the total.
+		// If a node is launched/registered, but not initialized, pods aren't scheduled
+		// to the node, and these are treated as unhealthy until they're cleaned up.
+		// This prevents odd roundup cases with percentages where replacement nodes that
+		// aren't initialized could be counted towards the total, resulting in more disruptions
+		// to active nodes than desired, where Karpenter should wait for these nodes to be
+		// healthy before continuing.
+		if !node.Managed() || !node.Initialized() {
+			continue
+		}
+
+		nodePool := node.Labels()[v1beta1.NodePoolLabelKey]
+		numNodes[nodePool]++
+
+		// If the node satisfies one of the following, we subtract it from the allowed disruptions.
+		// 1. Has a NotReady conditiion
+		// 2. Is marked as disrupting
+		if cond := nodeutils.GetCondition(node.Node, v1.NodeReady); cond.Status != v1.ConditionTrue || node.MarkedForDeletion() {
+			disrupting[nodePool]++
+		}
+	}
+	nodePoolList := &v1beta1.NodePoolList{}
+	if err := kubeClient.List(ctx, nodePoolList); err != nil {
+		return nil, fmt.Errorf("listing node pools, %w", err)
+	}
+	for _, nodePool := range nodePoolList.Items {
+		minDisruptionsByReason := nodePool.MustGetAllowedDisruptions(ctx, clk, numNodes[nodePool.Name])
+		allowedDisruptionsTotal := 0
+
+		disruptionBudgetMapping[nodePool.Name] = map[v1beta1.DisruptionReason]int{}
+		for reason, minDisruptions := range minDisruptionsByReason {
+			// Subtract the allowed number of disruptions from the number of already disrupting nodes.
+			// Floor the value since the number of disrupting nodes can exceed the number of allowed disruptions.
+			// Allowing this value to be negative breaks assumptions in the code used to calculate how many nodes can be disrupted.
+			allowedDisruptions := lo.Clamp(minDisruptions-disrupting[nodePool.Name], 0, math.MaxInt32)
+			disruptionBudgetMapping[nodePool.Name][reason] = allowedDisruptions
+
+			allowedDisruptionsTotal += allowedDisruptions
+			BudgetsAllowedDisruptionsGauge.With(map[string]string{
+				metrics.NodePoolLabel: nodePool.Name, metrics.ReasonLabel: string(reason),
+			}).Set(float64(allowedDisruptions))
```

Comment:
This seems wrong, in that we'd be setting allowedDisruptions cumulatively based on the ordering in which these reasons are considered. I think realistically, we should be setting the metric value based on reason, and not worry about aggregating.

Reply:
this sets allowedDisruptions not allowedDisruptionsTotal, bit confused here